### DISCUSSION TITLE
Event listener in Nuxt v3

### DISCUSSION THREAD

**Discussion by maxdzin** [nuxt/nuxt#15950] [Questions] [answered]
Greetings everyone!

In Nuxt v2 we can perform events emitting/listening by `this.$nuxt.$emit/on/off`.
Is there some similar alternative in Nuxt v3?

**Comment by dammy001** [comment]
I don't think there's because it's part of breaking changes in vue 3.

**Reply by maxdzin** [reply]
Yes, right. But it would be good to have some alternative to it.

**Reply by dammy001** [reply]
@pi0

**Reply by eMeRiKa13** [reply]
I'm interested by this question. I would like to know how to communicate events between components in Nuxt3.

**Reply by aligzl** [reply]
You can use provide/inject instead

**Comment by Zielgestalt** [comment, marked as answer]
Do you want something like an event bus? Josh Deltner talked about building an event bus plugin with mitt (https://www.npmjs.com/package/mitt) at the nuxtnation conference:
https://youtu.be/0z69RXxJpcU?t=690

You can rebuild something like that in Nuxt3. Create a plugin:

```
import { defineNuxtPlugin } from '#app'
import mitt from 'mitt'
const emitter = mitt()

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.provide('bus', {
    $on: emitter.on,
    $emit: emitter.emit,
  })
})
```

Then you can emit an event e.g. for opening a video:

`this.$bus.$emit('videoOpened', { video: videoObject })`

And then you can listen to the events. In my case I have a tracking plugin that listens to certain events:

```
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin(({ $bus }) => {
  $bus.$on('chatClicked', ({ location }) => {
    console.log('Klicked! ', location)
  })

  $bus.$on('videoOpened', ({ video }) => {
    console.log('Video ', video.video_id)
  })
})
```

**Reply by maxdzin** [reply]
Sorry for the late response. Yes. That's exactly how I managed it.
Thank you!

**Reply by abhitheawesomecoder** [reply]
Will this work from component to page emit also? or only page too component? i.e. parent to child only or child to parent also?

**Reply by Zielgestalt** [reply]
This works app wide. You can emit everywhere and listen everywhere.

**Reply by Boluwatife2904** [reply]
Hello, good afternoon. How would You call this using the script setup in Nuxt 3? Thanks.

**Comment by anirudhmaddy** [comment]
Working code with Vue3

Create Folder `plugin` > Create file `useEmitter.js`

```
import { defineNuxtPlugin } from '#app'
import mitt from 'mitt'
const emitter = mitt()

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.provide('bus', {
    $on: emitter.on,
    $emit: emitter.emit,
  })
})
```
Emit the event from a `page` to a `component`. Add below code inside `setup` function

```
const { $bus } = useNuxtApp()
$bus.$emit('clickEvent', { "page": "Demo Page" })
```

Component where you want to receive the event. Add below code inside `setup` function 

```
 const { $bus } = useNuxtApp()

 $bus.$on('clickEvent', (data) => {
     // do whatever you want with data 
 })
```

**Comment by F-NiX** [comment]
I wrote a plugin for this ... [nuxt-bus](https://www.npmjs.com/package/nuxt-bus)

**Reply by meleshkevich** [reply]
thank you! your solution is easy to implement, exactly what I was looking for!
